### ISSUE TITLE
[UX] Views UI: Machine name field loses the ability to be automatically populated after view (human) name validation error.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2519]
Steps to reproduce:

1. Go to the "Add view" page (`admin/structure/views/add`)...
2. Intentionally do not fill in a name or any other information...
3. Scroll to the bottom of the form and hit either one of the "Continue & configure" or "Save & exit" submit buttons -> you get validation errors:

![backdrop-views_ui-add_view_form-validation_errors_if_view_name_and_page_path_are_left_empty](https://cloud.githubusercontent.com/assets/2423362/22653047/8567639e-ecdc-11e6-9929-1ff0ec9c138d.png)

4. Start filling in the view name field. Notice how the page title and path fields get populated automatically, but the machine name is not:

![backdrop-views_ui-add_view_form-machine_name_not_populated_after_validation_errors](https://cloud.githubusercontent.com/assets/2423362/22653081/9414d5fc-ecdc-11e6-8836-37d3dec8c1b9.png)

...also notice how despite the fact that the path filed is required, there is no indicator for that (no red asterisk).

---
PR upcoming...

**Comment by Dinornis** [comment]
Thanks @klonos I just tested the PR and it seems to fix the machine name problem.
> ...also notice how despite the fact that the path filed is required, there is no indicator for that (no red asterisk).

Since the page path field gets auto populated, I don't think it is necessary for it to be highlighted.

**Comment by klonos** [comment]
> Since the page path field gets auto populated, I don't think it is necessary for it to be highlighted.

Yes, in theory this should be the case, but only with this specific form and specific use case. There is a much deeper issue here with how the Form API, `#states` and validation work and I think it has been mentioned before (if not here, then in the d.org issue queue for sure, but I don't have the respective issues handy right now). I tried to incorporate a change that fixes the issue in my PR, but I changed my mind and started drafting a separate issue. I will file that issue once I do a bit more research on the matter.

Back to this specific use case though, you'll notice that there are 3 validation error messages but only two fields are highlighted. This is a usability issue for people not familiar with this form (or Backdrop/Drupal in general).

**Comment by Al-Rozhkov** [comment]
Thank you for PR, @klonos.

I have tested the sandbox and everything work as expected. Can we mark this issue as RTBC or we need something more?

**Comment by klonos** [comment]
Thanx for testing and for catching that typo there @Al-Rozhkov 👍 ...fixed in the PR. Setting to RTBC and assigning to @serundeputy and @quicksketch for when either one of them has time to take a second look at the code and merge if they approve.

**Comment by serundeputy** [comment]
I'm not sure this technique is accounting for the transliteration.  I defer to @quicksketch.

**Comment by klonos** [comment]
Yep, transliteration works just fine:

![backdrop-issue2519-pr1785](https://cloud.githubusercontent.com/assets/2423362/25068395/84d09df2-22a5-11e7-981d-b8bff5375a46.gif)

...although just on this form (after validation error). If you try to do the same thing by manually clicking the "[edit]" link next to the machine name, then this whole js magic stops, but I think that this is by design: if you click on the edit machine name link, you are likely to be entering your own thing; you don't want Backdrop messing with that as you change the view name.

**Comment by klonos** [comment]
...minor thing that doesn't work (with the current PR, after the form has been submitted with validation errors) is that if you start hitting backspace in order to delete characters, the last remaining character on the machine name field is not deleted:

![backdrop-issue2519-pr1785-machine_name-last_remaining_character_not_deleted](https://cloud.githubusercontent.com/assets/2423362/25068469/696d71f4-22a8-11e7-83ac-f8ec2fdef3ef.gif)

This works in the form before submitting:

![backdrop-issue2519-machine_name-last_remaining_character_properly_deleted](https://cloud.githubusercontent.com/assets/2423362/25068472/718ebd48-22a8-11e7-97a7-21474f0aa91b.gif)

**Comment by klonos** [comment]
...tried but my JS is not strong enough. I can't figure out how to fix that trivial glitch with the last character not being deleted. @quicksketch??

**Comment by quicksketch** [comment]
@klonos The PR does fix some issues, but I think the problem being fixed is a general issue with Machine Names, not one specific to Views. I tried creating a content type at `/admin/structure/types/add` and found the exact same problem: Save the form without any making any changes. This reveals the "Machine name" field, which no longer auto-updates as the user types.

So perhaps rather than fixing this in views.js, this should be done in machine-name.js and fix the problem everywhere instead of just Views?

**Comment by klonos** [comment]
@quicksketch good catch 👍 ...I'm surprised it did not occur to me to test this. Not an expert in js (or coding in general), but I will take a look at this when I get a chance (most likely during the next weekend).

Looks like I need to figure out which .js file holds the magic for the content type machine name generation and adjust that too. I'm thinking that with Views being in contrib for D7, it made sense to have separate code handling the views_ui machine name magic. But now with Views in core for Backdrop, if the code is similar between handling machine names in views_ui, the content type add/edit form, and elsewhere, I then say we should consolidate all these in a system-wide .js file, so that next time we fix things (or break them 😄) in a single place.

**Comment by klonos** [comment]
https://github.com/backdrop/backdrop/blob/1.x/core/misc/machine-name.js is the file. It already has an appropriate, generic name and it's in a system-wide path (as in non-dependent on any module) 👍

**Comment by docwilmot** [comment]
Can we have an update on this?

**Comment by docwilmot** [comment]
At line 77 https://github.com/backdrop/backdrop/blob/1.x/core/misc/machine-name.js#L77
```
       // Skip processing upon a form validation error on the machine name.
       if ($target.hasClass('error')) {
         return;
       }
```
This prevents the JS processing when theres a form error. Not sure why we dont want processing after an error though. One minor issue is that youd still get two errors although only one field is visible:
>Name field is required.
Machine-readable name field is required.

**Comment by BWPanda** [comment]
Here's a [PR](https://github.com/backdrop/backdrop/pull/3480) based on @docwilmot's suggestion. He pointed out that we can simply remove those lines of code from the JS file and then the machine name would be treated as usual during form validation errors. The downside is that then there's an error displayed for the machine name field when that field is actually hidden.

So my PR removes those JS lines to make machine name fields work normally, even after errors. It also hide errors relating to the machine name field being required when its source field also has an error displayed. The logic here is that if a machine name is empty, it's likely because the source field is empty, so fixing the source field will fix the machine name field too. I.e. no need for two separate errors to be displayed since the fields are linked.

**Comment by BWPanda** [comment]
Screencast from PR sandbox:

![Uploading Peek 2021-01-05 12-51.gif…]()
